pkg/scheduler/scheduler.go

```diff
@@ -85,26 +85,55 @@ func NewScheduler(logC log.Logger) *Scheduler {
 	}
 }
 
-const rateLimit = 5 * time.Second
+const (
+	rateLimit  = 5 * time.Second
+	numWorkers = 3
+)
+
+func (scheduler *Scheduler) poolWorker(numWorkers int, tasks chan Task) {
+	for i := 0; i < numWorkers; i++ {
+		go func(workerID int) {
+			for task := range tasks {
+				scheduler.log.Debug().Int("worker", workerID).Msg("scheduler: starting task")
+
+				if err := task.DoWork(); err != nil {
+					scheduler.log.Error().Int("worker", workerID).Err(err).Msg("scheduler: error while executing task")
+				}
+
+				scheduler.log.Debug().Int("worker", workerID).Msg("scheduler: finished task")
+			}
+		}(i + 1)
+	}
+}
 
 func (scheduler *Scheduler) RunScheduler(ctx context.Context) {
 	throttle := time.NewTicker(rateLimit).C
+	tasksWorker := make(chan Task, numWorkers)
+
+	// start worker pool
+	go scheduler.poolWorker(numWorkers, tasksWorker)
 
 	go func() {
 		for {
 			select {
 			case <-ctx.Done():
+				close(tasksWorker)
 				close(scheduler.stopCh)
 
 				return
 			default:
-				task := scheduler.getTask()
-				if task != nil {
-					if err := task.DoWork(); err != nil {
-						scheduler.log.Error().Err(err).Msg("error while executing task")
+				i := 0
+				for i < numWorkers {
+					task := scheduler.getTask()
+					if task != nil {
+						// push tasks into worker pool
+						scheduler.log.Debug().Msg("scheduler: pushing task into worker pool")
```

Comment:
This has the problem that we no longer get FIFO semantics.

Reply:
Please elaborate on this, what are you referring to?

Reply:
is "enqueue task" better? or this is about implementation and not logging messaging?